src/vm/eventpipefile.cpp

```diff
@@ -92,8 +93,61 @@ void EventPipeFile::WriteEvent(EventPipeEventInstance &instance)
     }
     CONTRACTL_END;
 
+    // Check to see if we've seen this event type before.
+    // If not, then write the event metadata to the event stream first.
+    unsigned int metadataId = GetMetadataId(*instance.GetEvent());
+    if(metadataId == 0)
+    {
+        metadataId = GenerateMetadataId();
+
+        EventPipeEventInstance* pMetadataInstance = EventPipe::GetConfiguration()->BuildEventMetadataEvent(instance, metadataId);
+        
+        Handle(*pMetadataInstance, 0); // 0 breaks recursion and represents the metadata event.
+
+        SaveMetadataId(*instance.GetEvent(), metadataId);
+
+        delete[] (pMetadataInstance->GetData());
+        delete (pMetadataInstance);
+    }
+
+    Handle(instance, metadataId);
+}
+
+void EventPipeFile::WriteEnd()
+{
+    CONTRACTL
+    {
+        THROWS;
+        GC_NOTRIGGER;
+        MODE_ANY;
+    }
+    CONTRACTL_END;
+
+    m_pSerializer->WriteObject(m_pBlock); // we write current block to the disk, whether it's full or not
+
+    m_pBlock->Clear();
+
+    // "After the last EventBlock is emitted, the stream is ended by emitting a NullReference Tag which indicates that there are no more objects in the stream to read."
```

Comment:
Accidental quotes on the comment?

Reply:
It was a quote from the docs, I should have added link